src/amm.rs

```diff
@@ -45,36 +46,60 @@ macro_rules! to_u128 {
     };
 }
 
+#[derive(PartialEq)]
+#[derive(Debug)]
+pub enum MathError {
+    DenominatorIsZero,
+    ResultOverflow
+}
+
+
 /// Calculating spot price given reserve of selling asset and reserve of buying asset.
 /// Formula : BUY_RESERVE * AMOUNT / SELL_RESERVE
 ///
 /// - `sell_reserve` - reserve amount of selling asset
 /// - `buy_reserve` - reserve amount of buying asset
 /// - `amount` - amount
 ///
-/// Returns None in case of error
-pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Option<Balance> {
-    ensure!(sell_reserve != 0);
+/// Returns MathError in case of error
+pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Result<Balance, MathError> {
+    if sell_reserve == 0 {
+        return Err(DenominatorIsZero);
+    }
+
+    if amount == 0 || buy_reserve == 0 {
+        return Ok(0u128)
+    }
 
     let (amount_hp, buy_reserve_hp, sell_reserve_hp) = to_u256!(amount, buy_reserve, sell_reserve);
 
     let spot_price_hp = buy_reserve_hp
         .checked_mul(amount_hp)
         .expect("Cannot overflow")
-        .checked_div(sell_reserve_hp)?;
+        .checked_div(sell_reserve_hp)
+        .expect("Cannot panic as reserve cannot be 0");
+
+    let result = to_u128!(spot_price_hp);
+    if result.is_none() {
+        return Err(ResultOverflow);
+    }
 
-    to_u128!(spot_price_hp)
+    Ok(result.unwrap())
```

Comment:
Yes, something like that.

Instead of these last 5-6 lines, you can simplify like this :

to_u128!(spot_price_hp).ok_or(ResultOverflow)

and use that to return result directly.